src/Tracepoints.jl

```diff
@@ -0,0 +1,497 @@
+        end
+    end
+end
+function disable!(mod::Module)
+    mod_lock = mod.__tracepoints_lock__
+    tracepoints = mod.__tracepoints_specs__
+    lock(mod_lock) do
+        for spec_dict in values(tracepoints)
+            for spec in values(spec_dict)
+                adjust_semaphore!(spec.semaphore, -1)
+            end
+        end
+    end
+end
+function enable!(mod::Module, category::Symbol)
+    mod_lock = mod.__tracepoints_lock__
+    tracepoints = mod.__tracepoints_specs__
+    lock(mod_lock) do
+        for spec in values(tracepoints[category])
+            adjust_semaphore!(spec.semaphore, 1)
+        end
+    end
+end
+function disable!(mod::Module, category::Symbol)
+    mod_lock = mod.__tracepoints_lock__
+    tracepoints = mod.__tracepoints_specs__
+    lock(mod_lock) do
+        for spec in values(tracepoints[category])
+            adjust_semaphore!(spec.semaphore, -1)
+        end
+    end
+end
+function enable!(mod::Module, category::Symbol, kind::Symbol)
+    mod_lock = mod.__tracepoints_lock__
+    tracepoints = mod.__tracepoints_specs__
+    lock(mod_lock) do
+        adjust_semaphore!(tracepoints[category][kind].semaphore, 1)
+    end
+end
+function disable!(mod::Module, category::Symbol, kind::Symbol)
+    mod_lock = mod.__tracepoints_lock__
+    tracepoints = mod.__tracepoints_specs__
+    lock(mod_lock) do
+        adjust_semaphore!(tracepoints[category][kind].semaphore, -1)
+    end
+end
+
+@static if VERSION >= v"1.7-"
+    function adjust_semaphore!(semaphore::ShareableSemaphore, adj::Int)
+        @atomic semaphore.value += adj
+    end
+    function probe_enabled(spec::TracepointSpec)
+        semaphore = spec.semaphore
+        return (@atomic semaphore.value) > 0
+    end
+else
+    function adjust_semaphore!(semaphore::ShareableSemaphore, adj::Int)
+        Threads.atomic_add!(semaphore.value, adj)
+    end
+    probe_enabled(spec::TracepointSpec) = spec.semaphore.value[] > 0
+end
+function probe_maybe_trigger(spec::TracepointSpec, category::Symbol, kind::Symbol, lib_id::Int64, abi_type, arg, name)
+    if probe_enabled(spec) && spec.payload[] != C_NULL
+        probe_trigger(spec, category, kind, lib_id, abi_type, arg, name)
+    end
+end
+@generated function probe_trigger(spec::TracepointSpec, category::Symbol, kind::Symbol, lib_id::Int64, ::Val{abi_type}, arg, name) where {abi_type}
+    if abi_type != :Any
+        if abi_type == :Nothing
+            Targs = Expr(:tuple, :Symbol, :Symbol, :Int64, :Symbol, :Int)
+        else
+            Targs = Expr(:tuple, :Symbol, :Symbol, :Int64, :Symbol, abi_type, :Symbol)
+        end
+    else
+        Targs = Expr(:tuple, :Symbol, :Symbol, :Int64, :Symbol, :Any)
+    end
+    ex = Expr(:call, :ccall, :(spec.payload[]), :Int64, Targs)
+    push!(ex.args, :(category))
+    push!(ex.args, :(kind))
+    push!(ex.args, :(lib_id))
+    push!(ex.args, QuoteNode(abi_type))
+    if abi_type != :Any
+        if abi_type == :Nothing
+            push!(ex.args, :(0))
```

Comment:
Why a special ABI where we always pass 0?

Reply:
This is here because it makes it easier to write probe functions (at least in Julia), as I can specify it as `(..., abi_type, arg...)` and then slurp up the correct number of arguments (1 or 2) depending on `abi_type`. If I don't pass something, then I'd need to wrap up `abi_type` into the arguments being splatted in and then unwrap it, inspect it, and unwrap further. I could always define multiple probes, but it's just more annoying to have to do that just to support the `Nothing` ABI.